src/StreamingBodyParser/MultipartParser.php

```diff
@@ -0,0 +1,319 @@
+<?php
+
+namespace React\Http\StreamingBodyParser;
+
+use Evenement\EventEmitter;
+use Psr\Http\Message\RequestInterface;
+use React\Http\HttpBodyStream;
+use React\Http\UploadedFile;
+use React\Promise\CancellablePromiseInterface;
+use React\Promise\Deferred;
+use React\Stream\ThroughStream;
+
+final class MultipartParser extends EventEmitter
+{
+    /**
+     * @var string
+     */
+    protected $buffer = '';
+
+    /**
+     * @var string
+     */
+    protected $ending = '';
+
+    /**
+     * @var int
+     */
+    protected $endingSize = 0;
+
+    /**
+     * @var string
+     */
+    protected $boundary;
+
+    /**
+     * @var RequestInterface
+     */
+    protected $request;
+
+    /**
+     * @var HttpBodyStream
+     */
+    protected $body;
+
+    /**
+     * @var CancellablePromiseInterface
+     */
+    protected $promise;
+
+    /**
+     * @var callable
+     */
+    protected $onDataCallable;
+
+    public function __construct(RequestInterface $request)
+    {
+        $this->onDataCallable = array($this, 'onData');
+        $deferred = new Deferred(function () {
+            $this->body->removeListener('data', $this->onDataCallable);
+            $this->body->close();
+        });
+        $this->promise = $deferred->promise();
```

Comment:
Is the promise property really needed? Looks like it's only used in `MultipartParser::cancel()`. I think this is unecessary and 

```php
$this->body->removeListener('data', $this->onDataCallable);
$this->body->close();
``` 
can be just called directly there.